src/persist/src/storage.rs

```diff
@@ -330,6 +330,167 @@ impl From<(&str, &str)> for LockInfo {
     }
 }
 
+/// Abstraction over a key value store that provides a compare-and-set/conditional put primitive
+///
+/// "real world" realizations of this api could be etcd / dynamodb / any sql database with transactions?
+pub trait CasKeyVal {
+    /// Returns a reference to the value corresponding to the key.
+    fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error>;
+
+    /// List all of the keys in the map that currently have a value.
+    fn list_keys(&self) -> Result<Vec<String>, Error>;
+
+    /// Conditionally insert/update a key-value pair into the map, if the existing
+    /// value matches expected.
+    ///
+    /// Return true if the cas executed correctly, and false if the current value != expected.
+    ///
+    /// TODO: could also have the return value be the current value here. Doing a simpler thing
+    /// for prototyping.
+    fn cas(
+        &mut self,
+        key: &str,
+        value: Option<Vec<u8>>,
+        expected: Option<Vec<u8>>,
+    ) -> Result<bool, Error>;
+}
+
+/// Abstraction over a key value store that supports atomic inserts, and scans
+/// from a monotonically increasing index
+///
+/// "real world" realizations of this api could be kafka / redpanda.
+pub trait AtomicQueue {
+    /// Append data to the queue, and return the index it was appended at.
+    ///
+    /// Indexes increase monotonically.
+    fn append(&mut self, data: Vec<u8>) -> Result<u64, Error>;
+
+    /// Read all data stored at indexes >= index.
+    fn scan(&self, index: u64) -> Result<Vec<(u64, Vec<u8>)>, Error>;
+
+    /// Delete all data stored at indexes < index
+    fn truncate(&mut self, index: u64) -> Result<(), Error>;
+}
+
+/// WIP
+pub trait Consensus {
+    /// Return the most recently stored data
+    fn head(&self) -> Result<Option<(SeqNo, Vec<u8>)>, Error>;
+
+    /// Atomically insert data at seqno, iff the current HEAD of the data is at seqno - 1.
+    ///
+    /// Return true if the insert is successful (seqno - 1 was the previous version and this
+    /// call to cas wrote at seqno), and false otherwise.
+    fn cas(&mut self, seqno: SeqNo, data: Vec<u8>) -> Result<bool, Error>;
+
+    /// Read all data inserted with sequence numbers >= seqno
+    fn scan(&self, seqno: SeqNo) -> Result<Vec<(SeqNo, Vec<u8>)>, Error>;
```

Comment:
btw, I think for etcd, we can do better than implementing this via CasKeyVal because the v3 api exposes revisions and compaction, which would let us use a single key (and definitely avoid the permanent deletion tombstone potential problem)

I haven't looked at what dynamo gives us yet

Reply:
yup yup! this was more about fleshing out what minimum requirements we would need from the underlying service to do this at all

Reply:
and yeah dynamo supports a scan https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html and a conditional delete https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteItem.html to avoid acruing tombstones

Reply:
> and yeah dynamo supports a scan https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html and a conditional delete https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteItem.html to avoid acruing tombstones

Reply:
> and yeah dynamo supports a scan https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Scan.html and a conditional delete https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_DeleteItem.html to avoid acruing tombstones

making sure we're on the same page here... I'm talking about an _internal_ tombstone, not like one we explicitly wrote ourselves. aka when we delete something, do we eventually recover the space?